program-runtime/src/loaded_programs.rs

```diff
@@ -720,35 +720,43 @@ impl LoadedPrograms {
     /// Extracts a subset of the programs relevant to a transaction batch
     /// and returns which program accounts the accounts DB needs to load.
     pub fn extract<S: WorkingSlot>(
-        &self,
+        &mut self,
         working_slot: &S,
         keys: impl Iterator<Item = (Pubkey, (LoadedProgramMatchCriteria, u64))>,
     ) -> ExtractedPrograms {
-        let environments = self.get_environments_for_epoch(working_slot.current_epoch());
+        let environments = self
+            .get_environments_for_epoch(working_slot.current_epoch())
+            .clone();
+        let loaded = Arc::new(Mutex::new(LoadedProgramsForTxBatch {
+            entries: HashMap::new(),
+            slot: working_slot.current_slot(),
+            environments: environments.clone(),
+        }));
         let mut missing = Vec::new();
-        let mut unloaded = Vec::new();
         let found = keys
             .filter_map(|(key, (match_criteria, count))| {
-                if let Some(second_level) = self.entries.get(&key) {
-                    for entry in second_level.iter().rev() {
+                if let Some(second_level) = self.entries.get_mut(&key) {
+                    for entry in second_level.iter_mut().rev() {
                         let current_slot = working_slot.current_slot();
                         if entry.deployment_slot <= self.latest_root_slot
                             || entry.deployment_slot == current_slot
                             || working_slot.is_ancestor(entry.deployment_slot)
                         {
                             if current_slot >= entry.effective_slot {
-                                if !Self::is_entry_usable(entry, current_slot, &match_criteria) {
-                                    missing.push((key, count));
+                                if let LoadedProgramType::Loading(waiting_list) = &entry.program {
+                                    waiting_list.lock().unwrap().push(loaded.clone());
```

Comment:
I _think_ there's a race condition here. One thread could find the `Loading` entry for a program. But before it could lock the `waiting_list` and insert itself, the other thread calls `submit_cooperative_loading_task()`, which'll remove the entry from the cache.

Reply:
Nah, the entire global cache itself is locked in the Bank.
I only put `waiting_list` into a `Mutex` because Rust does not realize that there is a outer lock and only one thread can ever access it. I also though about alternatively using `Arc::get_mut()` as we never clone `Loading` entries in TX batches.

Reply:
Makes sense. We should write a comment in the code explaining this. It's getting fairly complicated to remember these fine details after a few years.